src/stan_math_backend/Transform_Mir.ml

```diff
@@ -467,8 +467,16 @@ let trans_prog (p : Program.Typed.t) =
   (* Eval indexed eigen types in UDF calls to prevent
      infinite template expansion if the call is recursive
 
+     Infinite expansion can happen only when the call graph is cyclic.
+     The strategy here is to build the call graph one edge at the time
+     and check if adding that edge creates a cycle. If it does, insert
+     an `eval()` to stop template expansion.
+
      All relevant function calls are recorded transitively in `callgraph`,
      meaning if `A` calls `B` and `B` calls `C` then `callgraph[A] = {B,C}`.
+     In the worst case every function calls every other, `callgraph` has
+     size O(n²) and this algorithm is O(n³) so it's important to include
```

Comment:
Thanks for these comments, I find them very helpful.

A nitpick:
```suggestion
     size O(n^2) and this algorithm is O(n^3) so it's important to include
```

Just to keep our file encoding simpler I think we should stick to ASCII